### PR TITLE
Snowbridge V2 - Rewards Pallet

### PR DESCRIPTION
**Pull request by claravanstaden**
# Description

Adds a rewards pallet, where rewards for relayers will be accumulated. More details in the Snowbridge v2 design: https://github.com/paritytech/polkadot-sdk/blob/master/bridges/snowbridge/docs/v2.md#relayer-rewards

## Integration

Since this is a new feature, so it does not affect integration.

## Review Notes

TODOs:
[ ] Add deposit from Inbound and Outbound queue.
[ ] Test XCM
[ ] More tests, including integration.

### CODE REVIEW

**Comment by bkontur** [comment]
We have a similar pallet here: https://github.com/paritytech/polkadot-sdk/tree/master/bridges/modules/relayers for registering relayers (deposits, slashing) and collecting rewards for relayer accounts.

I'm just curious if we could have a single generic pallet for relayers/rewards that could cover both the P/K bridge, the Eth bridge, or any other possible bridge. That would be cool. Essentially, we need to store the same things:
- Relayer account
- Reward identifier (for the P/K bridge, we use the kind of `lane_id` identifier for accumulating rewards; for ETH, I see it uses `message_id`)
- Reward (the P/K bridge uses the native token (DOT/KSM), while the ETH bridge uses WETH)

Some good abstraction over the **reward identifier** to support both, and using `xcm::Asset` for the reward could potentially cover everything.

**Comment by burdges** [comment]
As an aside, we do not have grandpa or beefy rewards planned for validators, but if we ever need them then they'd need a median computation, like https://github.com/polkadot-fellows/RFCs/pull/119 does.

File: bridges/snowbridge/pallets/rewards/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			);
+			let deposit: Asset = (reward_asset, value).into();
+			let beneficiary: Location =
+				Location::new(0, Parachain(T::AssetHubParaId::get().into()));
+
+			let xcm: Xcm<()> = vec![
+				DepositAsset { assets: Definite(deposit.into()), beneficiary },
+				SetTopic(message_id.into()),
+			]
+			.into();
```
Comment (yrong) [inline bridges/snowbridge/pallets/rewards/src/lib.rs:144]: Withdraw/Teleport + BuyExecution is required for executing any XCM.

**Comment by claravanstaden** [comment]
@bkontur that would definitely be cool, to reuse what is already there. I could see in the future how we also would register relayers and use other functionality in the pallet. Right now we're kind of focused on meeting a tight deadline for v2 and although reusing is a great idea, it usually takes a bit longer to get right since it needs to cater to multiple party's needs. So I think we are going to do the Snowbridge specific pallet (luckily it is very simple) and revisit using your team's pallet in the future.

**Comment by claravanstaden** [comment]
Closing for now, still too WIP and depends on other PRs. Moving the PR to my personal clone: https://github.com/claravanstaden/polkadot-sdk/pull/1

**Pull request closed without merging**